Add Payment Request button support for Woo blocks

ricardo opened the pull request:
Fixes #1416
Fixes #1544
Fixes #1677

#### Changes proposed in this Pull Request

- Add Payment Request blocks integration. The code is adapted from the Stripe Payment Gateway (1467-gh-woocommerce/woocommerce-gateway-stripe).
- Use same normalization functions and API in blocks and shortcode.
- Remove some leftover code from shortcode integration (e.g. `clearCart`, unnecessary i18n and nonce variables) and make shortcode integration more similar to the classic checkout by using the `WCPayAPI`.
- Fix 3D Secure / SCA confirmation with payment request buttons.

There are a few improvements left to do:
- Perhaps change folder structure, moving everything related to blocks into `client/blocks/`. - This requires changing the current credit card checkout as well.
- Add loading indicator to PRB when 3DS modal is open or intent is being validated. - Not really related to this PR, as it applies to the shortcode implementation too.

#### Testing instructions

- Run `npm run build:client`.
- Install Gutenberg and WooCommerce Blocks.
- Create a second cart page and add the cart block to it.
- Create a second checkout page and add the checkout block to it.
- Test using the payment request button in the product, cart and checkout pages, both shortcode and blocks.
- Verify that the button shows up as expected.
- Verify that the order is created as expected when the payment succeeds.

-------------------

- [x] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

asumaran commented:
@ricardo I've tested my PR https://github.com/Automattic/woocommerce-payments/pull/1642 on top of your branch and didn't see any issue. My PR affects the Product page only.

ricardo commented:
@asumaran Thanks for testing it. I was concerned the changes I made to the payment request button class could break your PR.

reykjalin reviewed (changes requested):
Just noticed that Google Pay does not have a `light-outline` theme; only `dark` and `light`:

| PRB | Network request |
| :---: | :---: |
| <img width="257" alt="image" src="https://user-images.githubusercontent.com/13835680/116430637-8cde2080-a836-11eb-9cf8-371bf61c820b.png"> | <img width="712" alt="image" src="https://user-images.githubusercontent.com/13835680/116430718-a0898700-a836-11eb-839e-33d276affcce.png"> |

I think the fix for this is basically just to map `light-outline` to `light`?

reykjalin reviewed (changes requested):
When you fail the 3DS authentication while paying with a PRB the error message isn't properly formatted:

<img width="259" alt="image" src="https://user-images.githubusercontent.com/13835680/116431331-37564380-a837-11eb-92ac-b3d44eecfb7a.png">

<img width="544" alt="image" src="https://user-images.githubusercontent.com/13835680/116431432-50f78b00-a837-11eb-872d-1a18286aae98.png">

reykjalin reviewed (changes requested):
This is all coming along nicely! I think there's just some minor comments on the code, and the issues I pointed out ^ above.

I'm going to do some more testing, especially to make sure the shortcode flow is still working as expected following the changes to the API, but I think this is all generally looking pretty good ☺️ 

I really like how you've been able to adapt the Stripe implementation to fit better into the WCPay codebase. Excellent work! 👏

reykjalin reviewed (changes requested):
**This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR**

When I tried to pay **through the credit card form** with a card that will be declined (`4000 0000 0000 0002`) I don't get the expected `Card declined` error, and instead get this generic message at the top of the checkout page:

| Error Message | Credit Card Form |
| :---: | :---: |
| <img width="832" alt="image" src="https://user-images.githubusercontent.com/13835680/116463974-e6584680-a85a-11eb-85a8-9f9431803fea.png"> | <img width="528" alt="image" src="https://user-images.githubusercontent.com/13835680/116464001-f1ab7200-a85a-11eb-82f3-157f030a1d2c.png"> |

The PRB shows the expected error message in the correct place.

reykjalin reviewed (changes requested):
It seems the [aforementioned issue with failing 3DS verification through a PRB](https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647223068) might be an issue that existed before this PR, since the product page shows the exact same result:

<img width="826" alt="image" src="https://user-images.githubusercontent.com/13835680/116464331-61216180-a85b-11eb-8dce-5f5c97927d45.png">

reykjalin reviewed (changes requested):
Related to the [`light-outline` button not showing up correctly](https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647217822) it works on the Product page, so that's definitely new with this PR.

https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647217822

ricardo commented:
> This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR

Please move to a new issue 👍 

I have seen this error before, but didn't connect the dots. Well observed though.

It's probably because the block credit card method is still using the generic Gutenberg checkout endpoint to create the order, and that's not throwing any specific response from Stripe.

ricardo commented on client/payment-request/event-handlers.js:
```diff
@@ -0,0 +1,72 @@
+
+	if ( 'success' === response.result ) {
+		try {
+			const confirmation = api.confirmIntent( response.redirect );
+			// We need to call `complete` outside of `completePayment` to close the dialog for 3DS.
+			event.complete( 'success' );
+
+			// `true` means there is no intent to confirm.
+			if ( true === confirmation ) {
```
This was taken from `confirm-card-payment.js` [here](https://github.com/Automattic/woocommerce-payments/blob/0ece4e0791971a08e622134ff60d0e417e083392/client/checkout/blocks/confirm-card-payment.js#L17), from the credit card checkout.

> Actually — after reading the comment — I'm thinking willNotConfirm might be a better name? Or intentWasRejected?

I didn't understand the reasoning behind calling it the opposite name. Can you explain?

It kinda makes sense to call it `confirmation` or `confirmed` because `api.confirmIntent` might return `true` at first, but then it can return a `confirmation` Promise, which then resolves to a redirect URL.

reykjalin replied:
> it can return a confirmation Promise, which then resolves to a redirect URL.

Ahhh, in that case I think we should keep the `true === confirmation` check. And it also makes sense to keep the `confirmation` variable name.

>I didn't understand the reasoning behind calling it the opposite name. Can you explain?

No 😅 

I don't know why it's called the opposite. Maybe the blame will reveal who worked on this last? I doubt they remember the reasoning though 😬

ricardo commented on client/payment-request/utils/normalize.js:
```diff
@@ -0,0 +1,113 @@
+				: false;
+		} )
+		.filter( Boolean );
+};
```
Yes, makes sense 👍

ricardo commented on client/payment-request/utils/utils.js:
```diff
@@ -0,0 +1,117 @@
+/* global wcpayPaymentRequestParams */
+
+/**
+ * Retrieves payment request data from global variable.
+ *
+ * @param {string} key The object property key.
+ * @return {mixed} Value of the object prop or null.
+ */
+export const getPaymentRequestData = ( key ) => {
+	if (
+		'object' === typeof wcpayPaymentRequestParams &&
+		wcpayPaymentRequestParams.hasOwnProperty( key )
+	) {
+		return wcpayPaymentRequestParams[ key ];
+	}
+	return null;
+};
```
I thought it would be "uglier" to add the wcpayPaymentRequestParams global variable in React everywhere, so I added this "getter function", so that we know where it's coming from. It's supposed to look similar to `getStripeServerData` from the Stripe implementation.

Do you think it's better to use `wcpayPaymentRequestParams?.key` along with `/* global wcpayPaymentRequestParams */` at the header instead? I'm happy to change.

reykjalin replied:
Ahhh, that makes sense actually; wrapping the global variable like that. I didn't connect the dots there.

I think I'd prefer to keep the function, provided that we make it clearer in the function body that `wcpayPaymentRequestParams` is a global? Possibly with a `global wcpayPaymentRequestParams` at the top of the function — if that doesn't break anything — or a comment?

ricardo replied:
There's a `/* global wcpayPaymentRequestParams */` comment at the top, where we usually place this sort of comment. Marking as resolved.

ricardo commented on client/payment-request/utils/normalize.js:
```diff
@@ -0,0 +1,113 @@
+ * @return {Object} Order object in the format WooCommerce expects.
+ */
+export const normalizeOrderData = ( paymentData ) => {
+	const name = paymentData.paymentMethod.billing_details.name;
+	const email = paymentData.paymentMethod.billing_details.email;
+	const phone = paymentData.paymentMethod.billing_details.phone;
+	const billing = paymentData.paymentMethod.billing_details.address;
+	const shipping = paymentData.shippingAddress ?? {};
```
It makes sense to play safe here. It even fixes this problem I was about to fix in a separate PR: #1677

Thanks!

ricardo commented:
> Just noticed that Google Pay does not have a light-outline theme; only dark and light:

Great catch! I changed the behavior a little bit, after looking at how Stripe renders the actual Google Pay button. The `light` version does not have a `box-shadow`.

I had to add a small vertical margin so that the ~~outline~~ shadow doesn't get cut by `overflow: hidden` in the blocks layout.

Fixed in [ffb0de4](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/ffb0de4a17cca2ba41cea09e35b9a0404cad50f4)

> It seems the aforementioned issue with failing 3DS verification through a PRB might be an issue that existed before this PR, since the product page shows the exact same result:

The issue was actually generated from this PR 😬 

Fixed it in [eb615a7](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/eb615a77c6b859d26e0ba97c78b7d4fba23e1f29).

reykjalin commented:
> > This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR
> 
> Please move to a new issue 👍

Issue opened in https://github.com/Automattic/woocommerce-payments/issues/1706

reykjalin commented on client/payment-request/style.scss:
```diff
@@ -11,26 +11,40 @@
 		&:focus {
-			box-shadow: #e8e8e8 0 1px 1px 0, #e8e8e8 0 1px 3px;
 			outline: 0;
 		}
```
We need to include the `box-shadow` here to indicate whether the element is focused when (for example) navigating via the <kbd>tab</kbd> key.

Personally, I think we should allow the purple outline set by WC Core, since that's really universal throughout WC pages; I don't see why we would remove that. Maybe @LevinMedia has some thoughts on that, design-wise?
The `box-shadow` on its own is **very difficult** to notice.

We should also make sure this is the same across both Stripe and WCPay, so what we decide to do here, I will also do in Stripe. For now I just added the `box-shadow` back in Stripe.

ricardo replied:
Good point about the accessibility issue.

I removed the `box-shadow` because the Google Pay button that Stripe renders (when you have the wallet available) doesn't have a `box-shadow`.

Apparently, it uses the same styling as `:hover`, so perhaps the right thing to do is to use:

```scss
&:hover, &:focus {
	background-color: #f8f8f8;
}
```

![image](https://user-images.githubusercontent.com/5509901/116766800-c19ed300-aa02-11eb-98ca-01099da38d31.png)

ricardo replied:
We can remove the `box-shadow` styling from `:focus` in the `light-outline` theme too. Stripe adds the same styling as in `:hover`.

![image](https://user-images.githubusercontent.com/5509901/116767011-e8a9d480-aa03-11eb-8bbe-cad9414bdced.png)

reykjalin replied:
The Stripe plugin version of this is not the best version of this either IMO, so I'd really like to get @LevinMedia's 👀 on this 🙂 

I'll ping him in Slack 👍

reykjalin commented on client/payment-request/style.scss:
```diff
@@ -46,13 +60,13 @@
 		&:focus {
-			box-shadow: #5f6368 0 1px 1px 0, #5f6368 0 1px 3px;
 			outline: 0;
 		}
```
Same as above ^, we need to include the `box-shadow` at the very least, and maybe allow the outline too.

ricardo replied:
Stripe also removes the outline property. I guess the best here is to be compliant with Stripe's standards. Using the same `background-color` as in `:hover` fixes this.

![image](https://user-images.githubusercontent.com/5509901/116767290-a386a200-aa05-11eb-9ea6-7fb5abcc9de8.png)

reykjalin reviewed (changes requested):
Just 2 minor comments on the updated PRB styling, but otherwise this looks good!

All the issues I encountered during my previous review have been fixed! Awesome work! 👏 

Let's get that CSS fixed (possibly with the help of @LevinMedia) and then get this shipped 🎉

reykjalin reviewed (approved):
This tests and works really well!

At least the most common use cases; I know you've been working on some edge cases, but the only one I (know of and) tested was where removing a product and leaving only a virtual product in the cart wasn't removing shipping options from the PRB.

I like the updated CSS for the PRBs too; you're right setting the background color to highlight it works well. I still think we should allow the purple outline, but that's not really up to me (cc @LevinMedia) 😛 

The code looks great too! I have one last minor suggestion, but that's not a blocker for merging 🙂 

I think this looks great, and it's time to ship! 🚀 
If there are more edge cases you want me to test, please let me know and I'll run through  them ☺️

ricardo commented on client/payment-request/blocks/use-initialization.js:
```diff
@@ -0,0 +1,160 @@
+		let shippingAddressChangeEvent = noop,
+			shippingOptionChangeEvent = noop,
+			paymentMethodChangeEvent = noop,
+			cancelChangeEvent = noop;
```
@reykjalin Fixed in [`bef480e`](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/bef480e613ee5603752fb9f07c6b26c486437c30).

I went a little further and removed the individual `removeAllListeners()` by using `paymentRequest.removeAllListeners()` at the top.

To test, I added a `console.log` to the event handlers to see if they are't being triggered more than once, and this seems to work well.

reykjalin replied:
> I went a little further and removed the individual `removeAllListeners()` by using `paymentRequest.removeAllListeners()` at the top.

I don't know how I feel about that tbh. The function returned from `useEffect` is intened to [clean up resources before the component leaves the screen](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect).

I don't think there's anything wrong with the approach in this particular case, but it's "bad form" so to speak. I think it'd be better to return a function that calls `paymentRequest.removeAllListeners()` 🤔 :

```js
// add handlers
const shippingAddressChangeHandler = paymentRequest?.on( '', () => {} );
const anotherHandler = paymentRequest?.on( // ...

return () => {
    paymentRequest?.removeAllListeners();
}
```

It feels a bit better to use the cleanup function for this purpose, but I'm not _strictly opposed_ to the current approach. I would say I'd strongly recommend going with the cleanup function calling `removeAllListeners()` 🙂

ricardo replied:
@reykjalin Fixed in [`5b2980c`](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/5b2980c33074965c9da6b9f48535aeede1c5ab6a), let me know how that looks 🙂

reykjalin replied:
Looks awesome! 🎉

LevinMedia commented:
@reykjalin @ricardo - let's just use the styling that stripe provides for now - I'm hesitant to use a specific color outline for focus since it may not always work with a theme's styles 👍

asumaran commented on client/payment-request/event-handlers.js:
```diff
@@ -0,0 +1,72 @@
+};
+
+export const shippingOptionChangeHandler = async ( api, event ) => {
+	const response = await api.paymentRequestUpdateShippingDetails( event );
```
@ricardo it should've been `event.shippingOption`. I'm submitting a PR fixing it.
